File: packages/ts/core/src/headers.ts
import type {
  ConditionalHeadersInput,
  ConditionalRequestHeaders,
  IdempotencyHeaders,
} from "./types.js";

export function conditionalHeaders(input: ConditionalHeadersInput): ConditionalRequestHeaders {
  const headers: ConditionalRequestHeaders = {};
  const ifMatch = input.ifMatch ?? input.etag;
  if (ifMatch) {
    headers["If-Match"] = ifMatch;
  }

  if (input.ifNoneMatch) {
    headers["If-None-Match"] = input.ifNoneMatch;
  }

  return headers;
}

export function idempotencyHeaders(key: string = crypto.randomUUID()): IdempotencyHeaders {
  return { "Idempotency-Key": key };
}

export function responseEtag(response: Response | undefined): string | undefined {
  return response?.headers.get("ETag") ?? undefined;
}

File: packages/python/core/sendmux_core/headers.py
from __future__ import annotations


def idempotency_headers(idempotency_key: str | None = None) -> dict[str, str]:
    return {"Idempotency-Key": idempotency_key} if idempotency_key else {}


def conditional_headers(*, if_match: str | None = None, if_none_match: str | None = None) -> dict[str, str]:
    headers: dict[str, str] = {}
    if if_match:
        headers["If-Match"] = if_match
    if if_none_match:
        headers["If-None-Match"] = if_none_match
    return headers

